app/javascript/packages/masked-text-toggle/index.js

```diff
@@ -0,0 +1,29 @@
+class MaskedTextToggle {
+  /**
+   * @param {HTMLInputElement} toggle
+   */
+  constructor(toggle) {
+    const controls = /** @type {string} */ (toggle.getAttribute('aria-controls'));
+    const textWrapper = /** @type {HTMLElement} */ (document.getElementById(controls));
+    const texts = /** @type {NodeListOf<HTMLElement>} */ (textWrapper.querySelectorAll(
+      '.masked-text__text',
+    ));
```

Comment:
I forget, do we have optional chaining? if so, would it make sense to do something like this?

```suggestion
    const texts = /** @type {NodeListOf<HTMLElement>} */ (
      document.getElementById(controls)?.querySelectorAll('.masked-text__text') || []
    );
```

And without optional chaining, should we check that `textWrapper` is not null still? Or are we safe to assume that the DOM will be shaped like we expect

Reply:
>I forget, do we have optional chaining?

We do in contexts like this where we're assuming Babel will transpile the code. It was an issue in #5223 because the Webpack plugin is run from source.

>if so, would it make sense to do something like this?

Maybe! The biggest caveat is that `NodeListOf` is not the same as an array (see [`NodeList`](https://developer.mozilla.org/en-US/docs/Web/API/NodeList)), which I suspect TypeScript will recognize and become upset about. Also, I personally find that `possiblyUndefined?.someValue || 'a default'` can be a little hard to follow that the `?.` is a trigger for the default, though I don't feel very strongly about it.

> And without optional chaining, should we check that `textWrapper` is not null still? Or are we safe to assume that the DOM will be shaped like we expect

I could go either way with this. Currently, the code definitely assumes that the elements exists and doesn't do any validation. It's the whole reason for the type-casting, essentially the same as native TypeScript's [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator). I felt mostly okay with this, at least in the context of knowing our app will satisfy those constraints. It might not pass muster if it were ever published as a reusable package, though.

In an earlier iteration, I did start with something that looked more like...

```js
const textWrapper = document.getElementById(controls);
if (!textWrapper) {
  throw new TypeError('missing text wrapper');
}
```

This has another upside in that the type-casting is no longer necessary, since TypeScript is smart enough to infer that any code past the condition can use the text wrapper as non-null.

I turned away from the validation mostly because it gets a bit verbose depending how far down the chain of dependencies we care to validate (e.g. does it have `aria-controls`? does the `aria-controls` target exist? etc.).

Reply:
> The biggest caveat is that `NodeListOf` is not the same as an array (see [`NodeList`](https://developer.mozilla.org/en-US/docs/Web/API/NodeList)),

We could probably accommodate this:

```js
const texts = /** @type {HTMLElement[]} */ (Array.from(
  document.getElementById(controls)?.querySelectorAll('.masked-text__text') || []
));
```

Reply:
Yeah I never know what the best way to keep the server-side template + JS in sync is, if it's better to just assume the pieces are there and error loudly or not. As long as we have a test that checks they're wired together its probably inconsequential

Reply:
Yeah, another angle we could take here is to move some of the DOM building into the JavaScript. It has the upside in that we only need to validate ideally a single "container" is given, and that the elements won't be shown in a no-JS context. But also downsides in that building DOM is clunky without a library (e.g. React) and there'd be some page shifting between initialization.

The tests do help us, but also it's another point of concern, since it's yet another place we have to duplicate the markup. If this were a JS-only project, this is something where an approach like [React's `hydrate`](https://reactjs.org/docs/react-dom.html#hydrate) can come in handy to keep everything in sync between server and client. I have seen some interesting / horrifying explorations of bridging this gap across languages which could be interesting to explore at some point (e.g. [React-PHP-V8Js](https://github.com/reactjs/react-php-v8js)).

Reply:
I went a slightly different route in ffd9ecb41, leveraging the fact that...

```js
document.getElementById('id')?.querySelectorAll('.class');
```

...is the same as:

```js
document.querySelectorAll('#id .class');
```